idx/memory/tag_query.go

```diff
@@ -50,36 +61,61 @@ func (a KvReByCost) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a KvReByCost) Less(i, j int) bool { return a[i].cost < a[j].cost }
 
 type TagQuery struct {
-	from      int64
-	equal     []kv
-	match     []kvRe
-	notEqual  []kv
-	notMatch  []kvRe
-	startWith int
+	from        int64
+	equal       []kv
+	match       []kvRe
+	notEqual    []kv
+	notMatch    []kvRe
+	prefix      []kv
+	startWith   int // expression type to generate the initial result set from
+	filterTagBy int // if tags should be filtered by prefix or regex match (max one)
+
+	// no need have more than one of tagMatch and tagPrefix, tags can only be
+	// filtered by max one condition
+	tagMatch  kvRe   // only used for /metrics/tags with regex in filter param
+	tagPrefix string // only used for auto complete of tags to match exact prefix
+
+	index TagIndex
+	byId  map[string]*idx.Archive
```

Comment:
when i see attributes for a type, i assume the attributes are scoped to the lifetime of the type. e.g. the fields are set when an instance is created, or very shortly after.
but index and byId are only initialized when Run() or RunGetTags() are called. this warrants a comment (or unless these fields can be set at creation time, that would be even a bit cleaner imho)

Reply:
The reason why they're only set after `Run()` or `RunGetTags()` got called is because we only acquire the index read lock after the query has been instantiated. That way, if the query is invalid, we can return an error without having to acquire the read lock. 
If we'd set the index properties to the query during instantiation, that would mean we'd set them without having the lock yet. This would work, but it would require that we first acquire the lock on the index before calling `Run()` / `RunGetTags()` and the index would not be an argument to those methods, so when reading the code that's using the queries it would be much less obvious how to make sure we are holding the proper locks. 
TL;DR this is really just to make it more obvious how to ensure that we lock/unlock the index at the right times.

Reply:
If we'd set the index during instantiation of the query we'd already need to acquire the read lock then for this bit: https://github.com/grafana/metrictank/blob/master/idx/memory/memory.go#L591
Currently when the query gets instantiated we verify that it is valid, if it is not valid we return an error without ever acquiring the index lock. That would not be possible then.

Reply:
that's fine. but let's add a comment than explaining whose responsibility is it to set those fields (seems Run and RunGetTags set it)